PWGHF/D2H/Tasks/taskCharmPolarisation.cxx

```diff
@@ -97,68 +112,138 @@ struct TaskPolarisationCharmHadrons {
+
+    // loop over mass hypotheses
+    for (uint8_t iMass = 0u; iMass < nMassHypos; iMass++) {
+
+      // variable definition
+      float pxDau{-1000.}, pyDau{-1000.}, pzDau{-1000.};
+      float pxCharmHad{-1000.}, pyCharmHad{-1000.}, pzCharmHad{-1000.};
+      float massDau{0.}, invMassCharmHad{0.}, invMassCharmHadForSparse{0.};
+      float rapidity{-999.};
+      std::array<float, 3> outputMl{};
+
+      if constexpr (channel == charm_polarisation::DecayChannel::DstarToDzeroPi) {
+        // Dstar analysis
+        // polarization measured from the soft-pion daughter (*)
+
+        pxDau = candidate.pxSoftPi();
+        pyDau = candidate.pySoftPi();
+        pzDau = candidate.pzSoftPi();
+        pxCharmHad = candidate.pxDstar();
+        pyCharmHad = candidate.pyDstar();
+        pzCharmHad = candidate.pzDstar();
+        massDau = massPi; // (*)
+        auto prongSoftPi = candidate.template prongPi_as<aod::Tracks>();
+        invMassCharmHad = (prongSoftPi.sign() > 0) ? candidate.invMassDstar() : candidate.invMassAntiDstar();
+        invMassCharmHadForSparse = (prongSoftPi.sign() > 0) ? (invMassCharmHad - candidate.invMassD0()) : (invMassCharmHad - candidate.invMassD0Bar()); // different for D*
+        rapidity = candidate.y(massDstar);
+        if constexpr (withMl) {
+          outputMl[0] = -1.; // not yet implemented in the selector
+          outputMl[1] = -1.; // not yet implemented in the selector
+          outputMl[2] = -1.; // not yet implemented in the selector
+        }
+      } else if constexpr (channel == charm_polarisation::DecayChannel::LcToPKPi) {
+        // Lc->pKpi analysis
+        // polarization measured from the proton daughter (*)
+
+        // reconstructed as pKpi
+        if (iMass == charm_polarisation::MassHyposLcToPKPi::PKPi && candidate.isSelLcToPKPi() >= selectionFlagLcToPKPi) {
+          pxDau = candidate.pxProng0();
+          pyDau = candidate.pyProng0();
+          pzDau = candidate.pzProng0();
+          invMassCharmHadForSparse = hfHelper.invMassLcToPKPi(candidate);
+        }
+
+        // reconstructed as piKp
+        if (iMass == charm_polarisation::MassHyposLcToPKPi::PiKP && candidate.isSelLcToPiKP() >= selectionFlagLcToPKPi) {
+          pxDau = candidate.pxProng2();
+          pyDau = candidate.pyProng2();
+          pzDau = candidate.pzProng2();
+          invMassCharmHadForSparse = hfHelper.invMassLcToPiKP(candidate);
+        }
+
+        // NB: no need to check cases in which candidate.isSelLcToPKPi() and candidate.isSelLcToPiKP() are both false, because they are rejected already by the Filter
+
+        // hypothesis-independent variables
+        pxCharmHad = candidate.px();
+        pyCharmHad = candidate.py();
+        pzCharmHad = candidate.pz();
+        massDau = massProton; // (*)
+        rapidity = candidate.y(massLc);
+        if constexpr (withMl) {
+          outputMl[0] = -1.;
+          outputMl[1] = -1.;
+          outputMl[2] = -1.;
+        }
```

Comment:
For the Lc->pKpi there is the implemented ML table, why don't you use it here?

Reply:
going to solve it. Apparently the Lc has 2 different sets of BDT scores for pKpi and piKp, I need to change the things slightly